### PR TITLE
Bag of Words, GAT changes, hate speech, and refactoring

### PR DESCRIPTION
**Pull request by jpngfile**
I tried using Bag of words as a text representation, made some improvements to the GAT model to match the paper, and refactored some methods into separate files. I'm still brainstorming how to improve the performance as it doesn't predict better than choosing one class on score polarity.

With our implementation of GAT, I was running into an issue where parameters were not changing with 2 layers of GAT. I found another python implementation of GAT on Github which is what I'm using now. I've included to this repo as a submodule (https://github.com/Diego999/pyGAT). 

## Submodules
- Added `pyGAT` as an implementation of GAT.
- Added `hate-speech-and-offensive-language` for its list of common hate phrases

To add the submodules, run:
```
git submodule init
git submodule update
```

## GAT Changes
- Added dropout to the model
- Changed leaky Relu to ELU

These changes were in the paper, and are also present in `pyGAT`

## text processing
- Added bag of words preprocessing
- Added filtering of stop words (e.g. remove words like "a", "the", "because", etc.)
- Added word stemming to avoid repeated words in bag of words
- Changed adjacency matrix so connections are undirected
- Added method for printing comments with hate speech
- Added method for filtering nodes with fewer than 4 edges

## Training
- Created file for testing Bag of words. One is with our implementation of GAT, and the other uses `pyGAT`. In the `pyGAT` case, it also only optimizes for nodes with at least 4 edges.

## Pip requirements
- Updated `requirements.txt` with more text processing packages. Run `pip install -r requirements.txt` to update

### CODE REVIEW

**Comment by agpar** [comment]
> With our implementation of GAT, I was running into an issue where parameters were not changing with 2 layers of GAT. 

Sounds like it must be some small configuration issue? If one layer is working and two is not then we must just not be composing them correctly... no?

**Comment by agpar** [comment]
> Changed adjacency matrix so connections are undirected.

This is an interesting change. Can you motivate it a bit?

**Comment by agpar** [comment]
I think based on this PR there are some interesting changes being made. Let me spell out how I approached this problem originally, and you can tell me if you see this direction as being not desirable or not doable with our current technology choices.

I just want to make sure we're on the same page w.r.t. where I was originally coming from, in case there is some misunderstanding regarding the larger vision.

* I want to figure out if a comment is engaging in some kind of bad behaviour. I don't think I can do this very reliably with NLP, since things like context and sarcasm make a lot of nasty behaviour difficult to detect directly from text.
* I hypothesize that since humans are good nasty behaviour detectors, instead of looking at some comment's text directly, I could look at how that comment was responded to (i.e. the entire sub tree rooted at that comment), and see if I can find an association between some kinds of easily detectable bad behaviour and these responses.
* To *really* test if this works, I decide to totally ignore any information about the comment at the root of the sub-tree, and *only* look at the responses to it.
* Based on this approach, only comments that actually created an "interesting" discussion can be targets of prediction. I decide this means if a comment has at least two children and at least one grand child, then I will call in "interesting", as I might have enough information in the descendants to say something about the parent.
    * Therefore, I don't bother trying to predict anything about comments that are leafs, or who aren't sufficiently interesting. (as defined above) A leaf comment can provide data to predict something about it's ancestor, but it itself is not a target for optimization or prediction.
   * Therefore, since I'm doing some lexical analysis, I only care about English comments.
   * Therefore, a single tree of comments can produce multiple data-label pairs, as any interior node of that tree that is interesting (as defined above) could be used as a target for prediction.

Moving to the GAT architecture, I imagined trying to predict a label for a single node (the parent) based on the descendants. This could either mean having the root node with no data and trying to classify it, or simply excluding the parent node and trying to classify the entire subtree (minus the root) as being the result of good or bad behaviour (in the root). Since the GAT architecture is for node classification, I thought it would be more like the former.

It looks like you're trying to predict on all nodes in the tree at once - I'm not sure how this effects this vision. Certainly it will make accuracy worse if you're trying to predict qualities of leaf nodes.
You've also changed the definition of "interesting" and removed the notion of directed adjacency. I'm wondering what inspired these changes?

**Comment by agpar** [comment]
I think that although the GAT architecture tries to classify every node, we could get around this by having our loss function only reflect loss on the label of the root node. This is just an idea.
This could either happen in a custom loss function or in the final layer of the network (a layer that just ignored every node prediction except the root).

**Comment by jpngfile** [comment]
For the 2 layer GAT, it could be a configuration issue, but I suspect it has to do with the implementation of combining the attention layers together. While debugging this, I found `pyGAT`, which worked so I just moved to using that.

I changed to undirected edges to follow the Cora dataset. I wanted to try new approaches to see if nodes didn't have enough neighbours to be able to effectively predicted. This could be changed back to directed edges.

I changed the definition of "interesting" to 4 edges to simplify filtering the nodes. Each node is it's own neighbour, and most will have a parent node, so it's for nodes with at least 2 children. I removed the grandchild requirement, since I was unsure if a reply to a reply would be relevant to the parent comment.

As for using GAT to predict interesting nodes, this is kind of an open question. I created the `FilterGAT` model which has a loss function only apply to "interesting" nodes, but I'm still investigating if this works. With two layers of GAT, the first layer needs to have an output for each node, which may add noise from the leaves.

I might move to just predicting parent comments with a single layer of GAT, so only direct children are considered.

**Comment by agpar** [comment]
Ok for switching to the library.

Ok, the adjacency thing can be considered basically a test.

I generally do think that the entire subtree can be interesting for classifying the root comment. The root "starts a discussion" and there could be patterns in the entire discussion (arguments, or mutual negativity, etc) that are an interesting signal about what the root started. I guess maybe we could say this effect gets smaller every jump you go away from the root, and the GAT approach will have to have some finite number of layers. Anyway, we can go with children only for now.
